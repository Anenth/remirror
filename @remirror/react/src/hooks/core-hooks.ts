import { SetStateAction, useCallback, useEffect, useLayoutEffect, useRef, useState } from 'react';
import ResizeObserver from 'resize-observer-polyfill';

import { isFunction, object } from '@remirror/core';

/**
 * Preserves the previous version of a provided value.
 *
 * ```tsx
 * const [isOpen, setOpen] = useState<boolean>(false)
 * const previous = usePrevious(isOpen)
 *
 * return <span onClick={() => setOpen(!isOpen)}>{isOpen && previous === isOpen ? 'Stable' : 'Unstable' }</span>
 * ```
 */
export function usePrevious<Value>(value: Value) {
  const ref = useRef<Value>();
  useEffect(() => void (ref.current = value), [value]);
  return ref.current;
}

export interface DOMRectReadOnlyLike {
  readonly x: number;
  readonly y: number;
  readonly width: number;
  readonly height: number;
  readonly top: number;
  readonly right: number;
  readonly bottom: number;
  readonly left: number;
}

const defaultBounds = { x: 0, y: 0, width: 0, height: 0, top: 0, right: 0, bottom: 0, left: 0 };

/**
 * Provides the measurements for a react element at the point of layout.
 *
 * @remarks
 *
 * Taken from https://codesandbox.io/embed/lp80n9z7v9
 *
 * ```tsx
 * const [bindRef, { height }] = useMeasure()
 *
 * return <div {...bindRef}>Height: {height}</div>
 * ```
 */
export function useMeasure<Ref extends HTMLElement = any>() {
  const ref = useRef<Ref>(null);
  const [bounds, setBounds] = useState<DOMRectReadOnlyLike>(defaultBounds);

  useLayoutEffect(() => {
    const resizeObserver = new ResizeObserver(([entry]) => setBounds(entry.contentRect));

    if (ref.current) {
      resizeObserver.observe(ref.current);
    }

    return () => resizeObserver.disconnect();
  }, []);

  return [{ ref }, bounds] as const;
}

export type DispatchWithCallback<Value> = (value: Value, callback?: () => void) => void;

interface UseStateWithCallback {
  <Type = undefined>(): readonly [
    Type | undefined,
    DispatchWithCallback<SetStateAction<Type | undefined>>,
  ];
  <Type>(value: Type): readonly [Type, DispatchWithCallback<SetStateAction<Type>>];
}

/**
 * Enables the use of state with an optional callback parameter in the setState
 * value.
 *
 * @remarks
 *
 * The callback is called once when the state next updates.
 */
export const useStateWithCallback: UseStateWithCallback = <State>(
  initialState?: State | (() => State),
) => {
  const [[state, callback], setState] = useState<[State | undefined, (() => void) | undefined]>([
    isFunction(initialState) ? initialState() : initialState,
    undefined,
  ]);

  useEffect(() => {
    if (callback) {
      callback();
      setState([state, undefined]);
    }
  }, [callback, state]);

  const setStateWithCallback = useCallback(
    (value: SetStateAction<State | undefined>, cb?: () => void) => {
      setState((prevState) => [isFunction(value) ? value(prevState[0]) : value, cb]);
    },
    [setState],
  );

  return [state, setStateWithCallback] as const;
};

export type PartialSetStateAction<State> = Partial<State> | ((prevState: State) => Partial<State>);

/**
 * A replication of the setState from class Components.
 *
 * @remarks
 *
 * It also accepts partial updates to the state object and a callback which
 * runs when the state has updated.
 *
 * It also returns a 3rd argument which resets the state to the original
 * initialState.
 *
 * ```ts
 * const [state, setState, resetState] = useSetState({a: 'initial', b: 'initial'});
 *
 * setState({a: 'A'});
 * log(state); // => { a: 'A', b: 'initial' }
 *
 * setState(prevState => ({b: 'B'}));
 * log(state); // => { a: 'A', b: 'B' }
 *
 * resetState();
 * log(state); // => { a: 'initial', b: 'initial' }
 * ```
 */
export function useSetState<State extends object>(
  initialState: State | (() => State) = object<State>(),
): readonly [
  State,
  DispatchWithCallback<PartialSetStateAction<State>>,
  (callback?: () => void) => void,
] {
  const [state, setStateWithCallback] = useStateWithCallback<State>(
    isFunction(initialState) ? initialState() : initialState,
  );

  const resetState = useCallback(
    (cb?: () => void) => {
      setStateWithCallback(initialState, cb);
    },
    [initialState, setStateWithCallback],
  );

  const setState = useCallback(
    (patch: PartialSetStateAction<State>, cb?: () => void) => {
      setStateWithCallback(
        (prevState: State) => ({
          ...prevState,
          ...(isFunction(patch) ? patch(prevState) : patch),
        }),
        cb,
      );
    },
    [setStateWithCallback],
  );

  return [state, setState, resetState] as const;
}
