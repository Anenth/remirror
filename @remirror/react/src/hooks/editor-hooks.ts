import { DependencyList, useContext, useMemo, useRef } from 'react';
import useDeepCompareEffect from 'use-deep-compare-effect';

import {
  AddCustomHandler,
  AddHandler,
  AnyCombinedUnion,
  AnyExtensionConstructor,
  AnyPresetConstructor,
  BuiltinPreset,
  Dispose,
  DynamicOptionsOfConstructor,
  EditorManager,
  ErrorConstant,
  ExtensionConstructorParameter,
  invariant,
  isFunction,
  keys,
  OptionsOfConstructor,
  PresetConstructorParameter,
} from '@remirror/core';
import { CorePreset } from '@remirror/preset-core';
import { ReactPreset } from '@remirror/preset-react';

import { RemirrorContext } from '../react-contexts';
import { RemirrorContextProps, UsePositionerParameter } from '../react-types';

/**
 * This provides access to the Remirror Editor context using hooks.
 *
 * @remarks
 *
 * The following example takes the position props from
 * ```ts
 * import { RemirrorProvider, useRemirror } from 'remirror';
 *
 * const HooksComponent = (props) => {
 *   // This pulls the remirror props out from the context.
 *   const { getPositionerProps } = useRemirror();
 *
 *   return <Menu {...getPositionerProps()} />;
 * }
 *
 * class App extends Component {
 *   render() {
 *     return (
 *       <RemirrorProvider>
 *         <HooksComponent />
 *       </RemirrorProvider>
 *     );
 *   }
 * }
 * ```
 */
export function useRemirror<Combined extends AnyCombinedUnion>(): RemirrorContextProps<Combined> {
  const context = useContext(RemirrorContext);

  invariant(context, { code: ErrorConstant.REACT_PROVIDER_CONTEXT });

  return context;
}

/**
 * A react hook for for creating an extension within the editor.
 *
 * @remarks
 *
 * This can be useful if you'd like to create your extensions and `EditorManager`
 * within a react component.
 *
 * @example
 *
 * ```ts
 * import { useCreateExtension, useCreatePreset } from 'remirror/react';
 * import { PresetCore } from 'remirror/preset-core';
 * import { BoldExtension } from 'remirror/extension-bold';
 *
 * const EditorWrapper = () => {
 *   const corePreset = useCreatePreset(PresetCore);
 *   const boldExtension = useCreateExtension(BoldExtension, { weight: 700 });
 *   const manager = useManager([corePreset, boldExtension]);
 *
 *   <RemirrorProvider >
 *     <MyEditor />
 *   </RemirrorProvider>;
 * }
 * ```
 *
 * You'll probably only need to work this way if you expect to dynamically
 * switch out the editor multiple times in apps like the
 * `@remirror/playground`.
 */
export function useCreateExtension<Type extends AnyExtensionConstructor>(
  Constructor: Type,
  ...[options]: ExtensionConstructorParameter<OptionsOfConstructor<Type>>
) {
  return useMemo(() => new Constructor(options), [Constructor, options]) as InstanceType<Type>;
}

/**
 * A react hook for for creating a preset within the editor.
 *
 * @remarks
 *
 * This can be useful if you'd like to create your presets and `EditorManager`
 * within a react component.
 *
 * @example
 *
 * ```ts
 * import { useCreateExtension, useCreatePreset } from 'remirror/react';
 * import { PresetCore } from 'remirror/preset-core';
 * import { BoldExtension } from 'remirror/extension-bold';
 *
 * const EditorWrapper = () => {
 *   const corePreset = useCreatePreset(PresetCore);
 *   const boldExtension = useCreateExtension(BoldExtension, { weight: 700 });
 *   const manager = useManager({extension: [boldExtension], presets: [] });
 *
 *   <RemirrorProvider >
 *     <MyEditor />
 *   </RemirrorProvider>;
 * }
 * ```
 *
 * You'll probably only need to work this way if you expect to dynamically
 * switch out the editor multiple times in apps like the
 * `@remirror/playground`.
 */
export function useCreatePreset<Type extends AnyPresetConstructor>(
  Constructor: Type,
  ...[settings]: PresetConstructorParameter<OptionsOfConstructor<Type>>
) {
  return useMemo(() => new Constructor(settings), [Constructor, settings]) as InstanceType<Type>;
}

/**
 * Dynamically update the properties of your extension via hooks. Provide the
 * Extension constructor and the properties you want to update.
 *
 * @remarks
 *
 * Please note that every time the properties change your extension is updated.
 * You will want to memoize or prevent needless updates somehow to the
 * properties passed in.
 *
 * This is only available within the context of the `RemirrorProvider` it will
 * throw an error otherwise.
 *
 * ```ts
 * import React, { useCallback, useState } from 'react';
 * import { useExtension } from 'remirror/react';
 * import { MentionExtension } from 'remirror/extension/mention';
 *
 * const FloatingQueryText = () => {
 *   const [query, setQuery] = useState('');
 *   const onChange = useCallback(({ query }) => setQuery(query.full), [setQuery]);
 *
 *   useExtension(MentionExtension, { appendText: 'hello' });
 *
 *   useExtension()
 *
 *   return <p>{query}</p>
 * }
 * ```
 *
 * The above example would add the `onChange` handler property to the extension.
 *
 * TODO - What about using this multiple times how could the extension handle
 * that?
 */
export function useExtension<Type extends AnyExtensionConstructor>(
  Constructor: Type,
  options: DynamicOptionsOfConstructor<Type>,
): void;
export function useExtension<Type extends AnyExtensionConstructor>(
  Constructor: Type,
  method: UseExtensionCallback<Type>,
  dependencies: DependencyList,
): void;
export function useExtension<Type extends AnyExtensionConstructor>(
  Constructor: Type,
  optionsOrCallback: DynamicOptionsOfConstructor<Type> | UseExtensionCallback<Type>,
  dependencies: DependencyList = [],
): void {
  const { manager } = useRemirror();

  const extension = useMemo(() => manager.getExtension(Constructor), [Constructor, manager]);

  // Handle the case where it an options object passed in.
  useDeepCompareEffect(() => {
    if (isFunction(optionsOrCallback)) {
      return;
    }

    extension.setOptions(optionsOrCallback);
  }, [extension, optionsOrCallback]);

  const memoizedDependencies = useObjectCheck(dependencies);

  useDeepCompareEffect(() => {
    if (!isFunction(optionsOrCallback)) {
      return;
    }

    return optionsOrCallback({
      addHandler: extension.addHandler,
      addCustomHandler: extension.addCustomHandler,
      extension,
    });
  }, [extension, optionsOrCallback, memoizedDependencies]);
}

interface UseExtensionCallbackParameter<Type extends AnyExtensionConstructor> {
  /**
   * Add a handler to the the extension callback.
   *
   * ```ts
   * addHandler('onChange', () => console.log('changed'));
   * ```
   */
  addHandler: AddHandler<OptionsOfConstructor<Type>>;

  /**
   * Set the value of a custom option which returns a dispose method. The custom
   * value is handled internally by the extension.
   *
   * ```ts
   * addCustomHandler('keybindings', { Enter: () => false });
   * ```
   */
  addCustomHandler: AddCustomHandler<OptionsOfConstructor<Type>>;

  /**
   * An instance of the extension. This can be used for more advanced scenarios.
   */
  extension: InstanceType<Type>;
}

type UseExtensionCallback<Type extends AnyExtensionConstructor> = (
  parameter: UseExtensionCallbackParameter<Type>,
) => Dispose;

/**
 * Update preset properties dynamically while the editor is still running.
 */
export function usePreset<Type extends AnyPresetConstructor>(
  Constructor: Type,
  options: DynamicOptionsOfConstructor<Type>,
): void;
export function usePreset<Type extends AnyPresetConstructor>(
  Constructor: Type,
  method: UsePresetCallback<Type>,
  dependencies: DependencyList,
): void;
export function usePreset<Type extends AnyPresetConstructor>(
  Constructor: Type,
  optionsOrCallback: DynamicOptionsOfConstructor<Type> | UsePresetCallback<Type>,
  dependencies: DependencyList = [],
): void {
  const { manager } = useRemirror();

  const preset = useMemo(() => manager.getPreset(Constructor), [Constructor, manager]);

  useDeepCompareEffect(() => {
    if (isFunction(optionsOrCallback)) {
      return;
    }

    preset.setOptions(optionsOrCallback);
  }, [preset, optionsOrCallback]);

  const memoizedDependencies = useObjectCheck(dependencies);

  useDeepCompareEffect(() => {
    if (!isFunction(optionsOrCallback)) {
      return;
    }

    return optionsOrCallback({
      addHandler: preset.addHandler,
      addCustomHandler: preset.addCustomHandler,
      preset,
    });
  }, [preset, optionsOrCallback, memoizedDependencies]);
}

function arePropertiesEqual<Type extends object>(value: Type, compare?: Type) {
  if (compare === undefined) {
    return false;
  }

  return keys(value).every((key) => compare[key]);
}

interface UsePresetCallbackParameter<Type extends AnyPresetConstructor> {
  /**
   * Add a handler to the the preset callback.
   *
   * ```ts
   * addHandler('onChange', () => console.log('changed'));
   * ```
   */
  addHandler: AddHandler<OptionsOfConstructor<Type>>;

  /**
   * Set the value of a custom option which returns a dispose method.
   */
  addCustomHandler: AddCustomHandler<OptionsOfConstructor<Type>>;

  /**
   * An instance of the preset. This should only be needed in advanced situations.
   */
  preset: InstanceType<Type>;
}

type UsePresetCallback<Type extends AnyPresetConstructor> = (
  parameter: UsePresetCallbackParameter<Type>,
) => Dispose;

/**
 * Takes an object and checks if any of it's properties are different from previously
 */
function useObjectCheck<Type extends object>(parameter: Type): Type {
  const ref = useRef<Type>();

  if (!ref.current || !arePropertiesEqual(parameter, ref.current)) {
    ref.current = parameter;
  }

  return ref.current;
}

/**
 * A hook for creating the editor manager directly in the react component.
 *
 * @remarks
 *
 * This might be helpful when creating tools like the `@remirror/playground`
 * which need to rerender the editor with different settings and can also
 * manager errors.
 *
 * ```tsx
 * import { useExtension } from '@remirror/react';
 * import { PresetCore } from '@remirror/preset-core';
 * import { BoldExtension } from '@remirror/extension-bold';
 *
 * const EditorWrapper = () => {
 *   const corePreset = usePreset(PresetCore);
 *   const boldExtension = useExtension(BoldExtension);
 *   const manager = useManager({ extension: [boldExtension], presets: [] });
 *
 *   <RemirrorProvider >
 *     <MyEditor />
 *   </RemirrorProvider>;
 * }
 * ```
 */
export function useManager<Combined extends AnyCombinedUnion>(
  combined: Combined[],
  settings: Remirror.ManagerSettings = {},
): EditorManager<Combined | ReactPreset | CorePreset | BuiltinPreset> {
  const extensionsAndPresets = useMemo(() => [...combined, new CorePreset(), new ReactPreset()], [
    combined,
  ]);

  return useMemo(() => EditorManager.fromList(extensionsAndPresets, settings), [
    extensionsAndPresets,
    settings,
  ]);
}

export type BaseReactCombinedUnion = ReactPreset | CorePreset | BuiltinPreset;

/**
 * A shorthand tool for retrieving the positioner props and adding them to a component.
 *
 * @remarks
 *
 * ```ts
 * import { bubblePositioner } from '@remirror/react';
 *
 * const MenuComponent: FC = () => {
 *   const {
 *     isActive,
 *     bottom,
 *     left
 * } = usePositioner({positionerId: 'bubbleMenu', positioner: bubblePositioner });
 *
 *   return (
 *     <div style={{ bottom, left }}>
 *       <MenuIcon {...options} />
 *     </div>
 *   );
 * }
 *
 * <RemirrorProvider extensions={[]}>
 *   <MenuComponent />
 * </RemirrorProvider>
 * ```
 */
export function usePositioner<Ref extends string = 'ref'>({
  positioner,
  ...rest
}: UsePositionerParameter<Ref>) {
  const { getPositionerProps } = useRemirror();

  return getPositionerProps<Ref>({ ...positioner, ...rest });
}
