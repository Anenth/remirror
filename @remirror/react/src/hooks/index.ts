export * from './core-hooks';
export * from './editor-hooks';
