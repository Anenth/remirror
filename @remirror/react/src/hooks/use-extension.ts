import { useEffect, useMemo, useRef } from 'react';

import {
  AnyExtension,
  AnyExtensionConstructor,
  ExtensionFromConstructor,
  GetProperties,
  GetSettings,
  IfNoRequiredProperties,
} from '@remirror/core';

import { useRemirror } from './use-remirror';

/**
 * Settings Of Extension Constructor
 *
 * Written in shorthand as it's only used in this file.
 */
type SettingsOfConstructor<Constructor extends AnyExtensionConstructor> = GetSettings<
  ExtensionFromConstructor<Constructor>
>;

export const useExtension = <Type extends AnyExtensionConstructor>(
  Constructor: Type,
  ...[settings]: IfNoRequiredProperties<
    SettingsOfConstructor<Type>,
    [SettingsOfConstructor<Type>?],
    [SettingsOfConstructor<Type>]
  >
) => {
  return useMemo(() => Constructor.of(settings), [Constructor, settings]);
};

/**
 * Properties Of Extension Constructor
 *
 * Written in shorthand as it's only used in this file.
 */
type PropertiesOfConstructor<Constructor extends AnyExtensionConstructor> = GetProperties<
  ExtensionFromConstructor<Constructor>
>;

export const useExtensionProperties = <Type extends AnyExtensionConstructor>(
  Constructor: Type,
  properties: PropertiesOfConstructor<Type>,
) => {
  const dependencyArray = useRef(Constructor.propertyKeys).current;
  const { manager } = useRemirror();

  const extension: AnyExtension = useMemo(() => manager.getExtension(Constructor), [
    Constructor,
    manager,
  ]);

  useEffect(() => {
    extension.setProperties(properties);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [extension, ...dependencyArray]);
};
