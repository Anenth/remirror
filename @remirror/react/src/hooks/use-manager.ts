import { useMemo, useRef } from 'react';

import {
  AnyExtension,
  AnyPreset,
  EditorManager,
  EditorManagerParameter,
  keys,
} from '@remirror/core';
import { CorePreset } from '@remirror/preset-core';
import { ReactPreset } from '@remirror/preset-react';

function arePropertiesEqual<Type extends object>(value: Type, compare?: Type) {
  if (compare === undefined) {
    return false;
  }

  return keys(value).every((key) => compare[key]);
}

/**
 * Takes an object and checks that each of it's keys are different
 */
function useObjectCheck<Type extends object>(parameter: Type): Type {
  const ref = useRef<Type>();

  if (!ref.current || !arePropertiesEqual(parameter, ref.current)) {
    ref.current = parameter;
  }

  return ref.current;
}

/**
 * A hook for creating the editor manager directly in the react component.
 *
 * @remarks
 *
 * This might be helpful when creating tools like the `@remirror/playground`
 * which need to rerender the editor with different settings and can also
 * manager errors.
 *
 * ```tsx
 * import { useExtension } from '@remirror/react';
 * import { PresetCore } from '@remirror/preset-core';
 * import { BoldExtension } from '@remirror/extension-bold';
 *
 * const EditorWrapper = () => {
 *   const corePreset = usePreset(PresetCore);
 *   const boldExtension = useExtension(BoldExtension);
 *   const manager = useManager([corePreset, boldExtension]);
 *
 *   <RemirrorProvider >
 *     <MyEditor />
 *   </RemirrorProvider>;
 * }
 * ```
 */
export function useManager<ExtensionUnion extends AnyExtension, PresetUnion extends AnyPreset>(
  parameter: EditorManagerParameter<ExtensionUnion, PresetUnion>,
) {
  const { extensions } = parameter;
  const settings = useObjectCheck(parameter.settings ?? {});
  const presets = useMemo(() => [...parameter.presets, new CorePreset(), new ReactPreset()], [
    parameter.presets,
  ]);

  return useMemo(() => EditorManager.create({ extensions, settings, presets }), [
    extensions,
    settings,
    presets,
  ]);
}
