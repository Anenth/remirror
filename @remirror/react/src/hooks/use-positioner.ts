import { UsePositionerParams } from '../react-types';
import { useRemirror } from './use-remirror';

/**
 * A shorthand tool for retrieving the positioner props and adding them to a component.
 *
 * @remarks
 *
 * ```ts
 * import { bubblePositioner } from '@remirror/react';
 *
 * const MenuComponent: FC = () => {
 *   const { isActive, bottom, left } = usePositioner({positionerId: 'bubbleMenu', positioner: bubblePositioner })
 *
 *   return (
 *     <div style={{ bottom, left }}>
 *       <MenuIcon {...properties} />
 *     </div>
 *   );
 * }
 *
 * <RemirrorProvider extensions={[]}>
 *   <MenuComponent />
 * </RemirrorProvider>
 * ```
 *
 * @param params
 * @param params.positioner
 * @param params.positionerId
 * @param params.refKey
 */
export const usePositioner = <GRefKey extends string = 'ref'>({
  positioner,
  ...rest
}: UsePositionerParams<GRefKey>) => {
  const { getPositionerProps } = useRemirror();

  return getPositionerProps<GRefKey>({ ...positioner, ...rest });
};
