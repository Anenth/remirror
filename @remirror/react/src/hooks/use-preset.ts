import { useEffect, useMemo } from 'react';

import {
  AnyPreset,
  AnyPresetConstructor,
  GetProperties,
  GetSettings,
  IfNoRequiredProperties,
  Of,
} from '@remirror/core';

import { useRemirror } from './use-remirror';

/**
 * Settings Of Extension Constructor
 *
 * Written in shorthand as it's only used in this file.
 */
type SettingsOfConstructor<Constructor extends AnyPresetConstructor> = GetSettings<Of<Constructor>>;

export const usePreset = <Type extends AnyPresetConstructor>(
  Constructor: Type,
  ...[settings]: IfNoRequiredProperties<
    SettingsOfConstructor<Type>,
    [SettingsOfConstructor<Type>?],
    [SettingsOfConstructor<Type>]
  >
) => {
  return useMemo(() => Constructor.of(settings), [Constructor, settings]);
};

/**
 * Properties Of Extension Constructor
 */
type PropertiesOfConstructor<Constructor extends AnyPresetConstructor> = GetProperties<
  Of<Constructor>
>;

/**
 * Update preset properties dynamically why the editor is still running.
 */
export const usePresetProperties = <Type extends AnyPresetConstructor>(
  Constructor: Type,
  properties: PropertiesOfConstructor<Type>,
) => {
  const { manager } = useRemirror();

  const preset: AnyPreset = useMemo(() => manager.getExtension(Constructor), [
    Constructor,
    manager,
  ]);

  useEffect(() => {
    preset.setProperties(properties);
  }, [preset, properties]);
};
