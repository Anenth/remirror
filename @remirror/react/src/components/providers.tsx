import React, { Children, ProviderProps } from 'react';

import { MakeOptional } from '@remirror/core';
import { InjectedRemirrorProps, RemirrorElementType, RemirrorFC, RemirrorProps } from '@remirror/react-utils';
import { defaultProps } from '../constants';
import { RemirrorContext } from '../contexts';
import { useRemirrorManager } from '../hooks';
import { Remirror } from './remirror';

export interface RemirrorContextProviderProps extends ProviderProps<InjectedRemirrorProps> {
  /**
   * Whether to set the first child as a the root prop (where the editor is rendered).
   *
   * If this is set to true then the outer element must be able to receive a ref prop which will mount
   * the editor to it. If not set then the children are responsible for calling `getRootProps`.
   *
   * @default false
   */
  setChildAsRoot?: boolean;
}

export type RemirrorProviderProps = MakeOptional<Omit<RemirrorProps, 'children'>, keyof typeof defaultProps> &
  Pick<RemirrorContextProviderProps, 'setChildAsRoot'>;

/**
 * This purely exists so that we know when the remirror editor has been called with a provider as opposed
 * to directly as a render prop by the user.
 *
 * It's important because when called directly by the user `getRootProps` is automatically called when the render prop
 * is called. However when called via a Provider the render prop renders the context component and it's not until
 * the element is actually rendered that the getRootProp in any nested components is called.
 */
const RemirrorContextProvider: RemirrorFC<RemirrorContextProviderProps> = ({
  setChildAsRoot: _,
  ...props
}) => {
  return <RemirrorContext.Provider {...props} />;
};

RemirrorContextProvider.$$remirrorType = RemirrorElementType.ContextProvider;
RemirrorContextProvider.defaultProps = {
  setChildAsRoot: false,
};

/**
 * The RemirrorProvider which injects context into it's child component.
 *
 * @remarks
 * This only supports one child. At the moment if that that child is an built in html string element
 * then it is also treated as the
 *
 * These can either be consumed using React Hooks
 * - `useRemirrorContext`
 * - `usePositioner`
 *
 * Or the higher order components
 * - `withRemirror`
 * - `withPositioner`
 */
export const RemirrorProvider: RemirrorFC<RemirrorProviderProps> = ({
  children,
  setChildAsRoot,
  ...props
}) => {
  return (
    <Remirror {...props}>
      {value => {
        return (
          <RemirrorContextProvider value={value} setChildAsRoot={setChildAsRoot}>
            {Children.only(children)}
          </RemirrorContextProvider>
        );
      }}
    </Remirror>
  );
};

RemirrorProvider.$$remirrorType = RemirrorElementType.EditorProvider;

export type ManagedRemirrorProviderProps = Omit<RemirrorProviderProps, 'manager'>;

/**
 * Renders the content while pulling the manager from the context and passing it on to the
 * RemirrorProvider.
 *
 * If no manager exists the child components are not rendered.
 */
export const ManagedRemirrorProvider: RemirrorFC<ManagedRemirrorProviderProps> = ({ children, ...props }) => {
  const manager = useRemirrorManager();

  return manager ? (
    <RemirrorProvider {...props} manager={manager}>
      {children}
    </RemirrorProvider>
  ) : null;
};

ManagedRemirrorProvider.$$remirrorType = RemirrorElementType.ManagedEditorProvider;
