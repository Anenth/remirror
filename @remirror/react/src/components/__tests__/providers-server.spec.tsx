/**
 * @jest-environment node
 */

import React from 'react';
import { renderToString } from 'react-dom/server';

import { docNodeBasicJSON } from '@test-fixtures/object-nodes';
import { createTestManager } from '@test-fixtures/schema-helpers';
import { useRemirror } from '../../hooks';
import { ManagedRemirrorProvider, RemirrorProvider } from '../providers';
import { RemirrorManager } from '../remirror-manager';

test('RemirrorProvider', () => {
  const TestComponent = () => {
    return <div data-testid='target' />;
  };

  const reactString = renderToString(
    <RemirrorProvider initialContent={docNodeBasicJSON} manager={createTestManager()}>
      <TestComponent />
    </RemirrorProvider>,
  );
  expect(reactString).toInclude('basic');
});

test('ManagedRemirrorProvider', () => {
  const TestComponent = () => {
    const { getRootProps } = useRemirror();

    return (
      <div>
        <div data-testid='target' {...getRootProps()} />
      </div>
    );
  };

  const reactString = renderToString(
    <RemirrorManager>
      <ManagedRemirrorProvider initialContent={docNodeBasicJSON}>
        <TestComponent />
      </ManagedRemirrorProvider>
    </RemirrorManager>,
  );
  expect(reactString).toInclude('basic');
});
