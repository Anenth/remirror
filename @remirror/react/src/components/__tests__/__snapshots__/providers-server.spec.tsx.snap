// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemirrorProvider 1`] = `"<div data-testid=\\"1\\"><div data-testid=\\"2\\"><style data-emotion-css=\\"186pr8o-css\\">.css-186pr8o-css .remirror-editor{caret-color:currentColor;word-wrap:break-word;white-space:pre-wrap;overflow-wrap:break-word;}.css-186pr8o-css .remirror-editor:focus{outline:none;}.css-186pr8o-css .is-empty:first-of-type::before{position:absolute;color:#aaa;pointer-events:none;height:0;font-style:italic;content:attr(data-placeholder);}.css-186pr8o-css .remirror-editor blockquote{border-left:2px solid #ddd;margin-left:0;margin-right:0;padding-left:10px;font-style:italic;}.css-186pr8o-css .remirror-editor blockquote p{color:#888;}</style><div data-testid=\\"target\\" class=\\"css-186pr8o-css\\"><div role=\\"textbox\\" aria-multiline=\\"true\\" aria-label=\\"\\" class=\\"remirror-editor\\"><p>basic</p></div></div></div></div>"`;
